File: SRC/dsytrd_2stage.f
      SUBROUTINE DSYTRD_2STAGE( VECT, UPLO, N, A, LDA, D, E, TAU, HOUS2, LHOUS2, WORK, LWORK, INFO )

      IMPLICIT NONE

*  -- LAPACK computational routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      String             VECT, UPLO;
      int                N, LDA, LWORK, LHOUS2, INFO;
      // ..
      // .. Array Arguments ..
      double             D( * ), E( * );
      double             A( LDA, * ), TAU( * ), HOUS2( * ), WORK( * );
      // ..

*  =====================================================================
      // ..
      // .. Local Scalars ..
      bool               LQUERY, UPPER, WANTQ;
      int                KD, IB, LWMIN, LHMIN, LWRK, LDAB, WPOS, ABPOS;
      // ..
      // .. External Subroutines ..
      // EXTERNAL XERBLA, DSYTRD_SY2SB, DSYTRD_SB2ST
      // ..
      // .. External Functions ..
      bool               LSAME;
      int                ILAENV2STAGE;
      // EXTERNAL LSAME, ILAENV2STAGE
      // ..
      // .. Executable Statements ..

      // Test the input parameters

      INFO   = 0
      WANTQ  = LSAME( VECT, 'V' )
      UPPER  = LSAME( UPLO, 'U' )
      LQUERY = ( LWORK == -1 ) .OR. ( LHOUS2 == -1 )

      // Determine the block size, the workspace size and the hous size.

      KD     = ILAENV2STAGE( 1, 'DSYTRD_2STAGE', VECT, N, -1, -1, -1 )
      IB     = ILAENV2STAGE( 2, 'DSYTRD_2STAGE', VECT, N, KD, -1, -1 )
      if ( N == 0 ) {
         LHMIN = 1
         LWMIN = 1
      } else {
         LHMIN = ILAENV2STAGE( 3, 'DSYTRD_2STAGE', VECT, N, KD, IB, -1 )
         LWMIN = ILAENV2STAGE( 4, 'DSYTRD_2STAGE', VECT, N, KD, IB, -1 )
      }

      if ( .NOT.LSAME( VECT, 'N' ) ) {
         INFO = -1
      } else if ( .NOT.UPPER .AND. .NOT.LSAME( UPLO, 'L' ) ) {
         INFO = -2
      } else if ( N.LT.0 ) {
         INFO = -3
      } else if ( LDA.LT.MAX( 1, N ) ) {
         INFO = -5
      } else if ( LHOUS2.LT.LHMIN .AND. .NOT.LQUERY ) {
         INFO = -10
      } else if ( LWORK.LT.LWMIN .AND. .NOT.LQUERY ) {
         INFO = -12
      }

      if ( INFO == 0 ) {
         HOUS2( 1 ) = LHMIN
         WORK( 1 )  = LWMIN
      }

      if ( INFO != 0 ) {
         xerbla('DSYTRD_2STAGE', -INFO );
         RETURN
      } else if ( LQUERY ) {
         RETURN
      }

      // Quick return if possible

      if ( N == 0 ) {
         WORK( 1 ) = 1
         RETURN
      }

      // Determine pointer position

      LDAB  = KD+1
      LWRK  = LWORK-LDAB*N
      ABPOS = 1
      WPOS  = ABPOS + LDAB*N
      dsytrd_sy2sb(UPLO, N, KD, A, LDA, WORK( ABPOS ), LDAB, TAU, WORK( WPOS ), LWRK, INFO );
      if ( INFO != 0 ) {
         xerbla('DSYTRD_SY2SB', -INFO );
         RETURN
      }
      dsytrd_sb2st('Y', VECT, UPLO, N, KD, WORK( ABPOS ), LDAB, D, E, HOUS2, LHOUS2, WORK( WPOS ), LWRK, INFO );
      if ( INFO != 0 ) {
         xerbla('DSYTRD_SB2ST', -INFO );
         RETURN
      }


      WORK( 1 ) = LWMIN
      RETURN

      // End of DSYTRD_2STAGE

      }

File: lib/src/dsytrd_2stage.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/install/lsame.dart';
import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/dsytrd_sb2st.dart';
import 'package:dart_lapack/src/dsytrd_sy2sb.dart';
import 'package:dart_lapack/src/ilaenv2stage.dart';
import 'package:dart_lapack/src/matrix.dart';
import 'package:dart_lapack/src/xerbla.dart';

void dsytrd_2stage(
  final String VECT,
  final String UPLO,
  final int N,
  final Matrix<double> A_,
  final int LDA,
  final Array<double> D_,
  final Array<double> E_,
  final Array<double> TAU_,
  final Array<double> HOUS2_,
  final int LHOUS2,
  final Array<double> WORK_,
  final int LWORK,
  final Box<int> INFO,
) {
  final A = A_.having(ld: LDA);
  final D = D_.having();
  final E = E_.having();
  final TAU = TAU_.having();
  final HOUS2 = HOUS2_.having();
  final WORK = WORK_.having();
  bool LQUERY, UPPER
      // WANTQ
      ;
  int KD, IB, LWMIN, LHMIN, LWRK, LDAB, WPOS, ABPOS;

  // Test the input parameters

  INFO.value = 0;
  // WANTQ = lsame(VECT, 'V');
  UPPER = lsame(UPLO, 'U');
  LQUERY = (LWORK == -1) || (LHOUS2 == -1);

  // Determine the block size, the workspace size and the hous size.

  KD = ilaenv2stage(1, 'DSYTRD_2STAGE', VECT, N, -1, -1, -1);
  IB = ilaenv2stage(2, 'DSYTRD_2STAGE', VECT, N, KD, -1, -1);
  if (N == 0) {
    LHMIN = 1;
    LWMIN = 1;
  } else {
    LHMIN = ilaenv2stage(3, 'DSYTRD_2STAGE', VECT, N, KD, IB, -1);
    LWMIN = ilaenv2stage(4, 'DSYTRD_2STAGE', VECT, N, KD, IB, -1);
  }

  if (!lsame(VECT, 'N')) {
    INFO.value = -1;
  } else if (!UPPER && !lsame(UPLO, 'L')) {
    INFO.value = -2;
  } else if (N < 0) {
    INFO.value = -3;
  } else if (LDA < max(1, N)) {
    INFO.value = -5;
  } else if (LHOUS2 < LHMIN && !LQUERY) {
    INFO.value = -10;
  } else if (LWORK < LWMIN && !LQUERY) {
    INFO.value = -12;
  }

  if (INFO.value == 0) {
    HOUS2[1] = LHMIN.toDouble();
    WORK[1] = LWMIN.toDouble();
  }

  if (INFO.value != 0) {
    xerbla('DSYTRD_2STAGE', -INFO.value);
    return;
  } else if (LQUERY) {
    return;
  }

  // Quick return if possible

  if (N == 0) {
    WORK[1] = 1;
    return;
  }

  // Determine pointer position

  LDAB = KD + 1;
  LWRK = LWORK - LDAB * N;
  ABPOS = 1;
  WPOS = ABPOS + LDAB * N;
  dsytrd_sy2sb(UPLO, N, KD, A, LDA, WORK(ABPOS).asMatrix(LDAB), LDAB, TAU,
      WORK(WPOS), LWRK, INFO);
  if (INFO.value != 0) {
    xerbla('DSYTRD_SY2SB', -INFO.value);
    return;
  }
  dsytrd_sb2st('Y', VECT, UPLO, N, KD, WORK(ABPOS).asMatrix(LDAB), LDAB, D, E,
      HOUS2, LHOUS2, WORK(WPOS), LWRK, INFO);
  if (INFO.value != 0) {
    xerbla('DSYTRD_SB2ST', -INFO.value);
    return;
  }

  WORK[1] = LWMIN.toDouble();
}
